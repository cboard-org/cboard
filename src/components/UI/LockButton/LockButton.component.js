import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import LockOutlineIcon from '@material-ui/icons/LockOutline';
import LockOpenIcon from '@material-ui/icons/LockOpen';

import IconButton from '../IconButton';
import messages from './LockButton.messages';

class LockButton extends PureComponent {
  static propTypes = {
    intl: intlShape.isRequired,
    isLocked: PropTypes.bool,
    maxClicks: PropTypes.number,
    onNotify: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired
  };

  static defaultProps = {
    maxClicks: 4
  };

  clicks = 0;
  timeout = 0;

  clearClicksTimeout(ms) {
    clearTimeout(this.timeout);

    this.timeout = setTimeout(() => {
      this.clicks = 0;
    }, ms);
  }

  handleClick = () => {
    // TODO: refactor into smaller functions
    const { intl, maxClicks, isLocked, onNotify, onClick } = this.props;

    this.clicks = this.clicks + 1;

    if (!isLocked) {
      this.clicks = 0;
      onClick();
      return;
    }

    if (this.clicks === 2) {
      onNotify(
        `${maxClicks - this.clicks} ${intl.formatMessage(
          messages.clicksToUnlock
        )}`
      );
    }

    if (this.clicks === maxClicks) {
      this.clicks = 0;
      onClick();
      return;
    }

    this.clearClicksTimeout(5000);
  };

  render() {
    const { intl, isLocked } = this.props;

    return (
      <IconButton
        label={
          isLocked
            ? intl.formatMessage(messages.unlock)
            : intl.formatMessage(messages.lock)
        }
        onClick={this.handleClick}
      >
        {isLocked ? <LockOutlineIcon /> : <LockOpenIcon />}
      </IconButton>
    );
  }
}

export default injectIntl(LockButton);
