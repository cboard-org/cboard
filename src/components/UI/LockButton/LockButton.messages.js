import { defineMessages } from 'react-intl';

export default defineMessages({
  lock: {
    id: 'cboard.components.Board.lock',
    defaultMessage: 'Lock'
  },
  unlock: {
    id: 'cboard.components.Board.unlock',
    defaultMessage: 'Unlock'
  },
  clicksToUnlock: {
    id: 'cboard.components.Board.clicksToUnlock',
    defaultMessage: 'clicks to unlock'
  }
});
