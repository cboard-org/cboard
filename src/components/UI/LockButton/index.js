export { default } from './LockButton.component';
