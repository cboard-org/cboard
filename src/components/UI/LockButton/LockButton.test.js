import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';

import LockButton from './LockButton.component';

describe('LockButton tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<LockButton onClick={() => {}} onNotify={() => {}} />);
  });
});
