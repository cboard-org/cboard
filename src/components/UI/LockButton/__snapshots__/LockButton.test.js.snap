// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LockButton tests default renderer 1`] = `
<LockButton
  intl={
    Object {
      "formatMessage": [Function],
    }
  }
  maxClicks={4}
  onClick={[Function]}
  onNotify={[Function]}
/>
`;
